internal/hcsoci/network.go

```diff
@@ -68,3 +70,63 @@ func GetNamespaceEndpoints(ctx context.Context, netNS string) ([]*hns.HNSEndpoin
 	}
 	return endpoints, nil
 }
+
+// NetworkSetup is used to abstract away the details of setting up networking
+// for a pod.
+type NetworkSetup interface {
+	ConfigureNetworking(ctx context.Context, namespaceID string) error
+}
+
+// LocalNetworkSetup implements the NetworkSetup interface for configuring pod
+// networking. It will setup the pods network locally.
+type LocalNetworkSetup struct {
+	VM *uvm.UtilityVM
+}
+
+func (l *LocalNetworkSetup) ConfigureNetworking(ctx context.Context, namespaceID string) error {
+	endpoints, err := GetNamespaceEndpoints(ctx, namespaceID)
+	if err != nil {
+		return err
+	}
+	if err := l.VM.AddNetNS(ctx, namespaceID); err != nil {
+		return err
+	}
+	err = l.VM.AddEndpointsToNS(ctx, namespaceID, endpoints)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// ExternalNetworkSetup implements the NetworkSetup interface for configuring
+// pod networking. It will try and communicate with an external network configuration
+// proxy service to setup networking.
+type ExternalNetworkSetup struct {
+	VM     *uvm.UtilityVM
+	CaAddr string
+	PodID  string
+}
+
+func (e *ExternalNetworkSetup) ConfigureNetworking(ctx context.Context, namespaceID string) error {
+	if err := e.VM.AddNetNS(ctx, namespaceID); err != nil {
+		return err
+	}
+	client := e.VM.NCProxyClient()
+	if client == nil {
+		return fmt.Errorf("no ncproxy client for UVM `%s`", e.VM.ID())
+	}
+	registerReq := &ncproxyttrpc.RegisterComputeAgentRequest{
+		ContainerID:  e.PodID,
+		AgentAddress: e.CaAddr,
+	}
+	if _, err := client.RegisterComputeAgent(ctx, registerReq); err != nil {
+		return err
+	}
+	netReq := &ncproxyttrpc.ConfigureNetworkingInternalRequest{
```

Comment:
I feel like this is confusing naming. I think it would be fine to just name this `ConfigureNetworkingRequest`. Thoughts?

Reply:
I couldn't because there's another ConfigureNetworkRequest defined for nodenetsvc so grpc complains about multiple of the same definition.